### ISSUE TITLE
Update mnist example to support step-by-step learning and remove inversion of control

### ISSUE THREAD

**Issue by buriy** [Lasagne/Lasagne#322]
I'd like to improve mnist example to avoid unnecessary complication and allow easier reusability.

Currently we do:
```python
def train(iter_funcs, dataset, batch_size=BATCH_SIZE):
    for epoch in itertools.count(1):
        ..do training...
        yield result

for epoch in train(iter_funcs, dataset):
    print epoch
```
Instead I'd like to see:
```python
def train_epoch(iter_funcs, dataset, batch_size=BATCH_SIZE):
    ..do training...
    return result

for epoch in itertools.count(1):
    result = train_epoch(iter_funcs, dataset)
    print result
```
The current way of doing it looks like an unnecessary complication to me, especially for the people who want to feed network with a lot of data in macro batches that don't fit into memory.

I think the library-like continuous learning concept
```python
some_more_data, some_more_labels = grab_more_data_and_labels()
network.learn(some_more_data, some_more_labels)
```
is more general and scalable, compared to framework-like learning-once concept:
```
network.learn_until_done(all_data, all_labels)
```

**Comment by ebenolson** [comment]
There's been discussion of this in https://github.com/Lasagne/Lasagne/issues/215, but I don't think anyone's working on it at the moment, so feel free to open a PR.

We're also starting to populate the [Recipes](https://github.com/Lasagne/Recipes) repo with other examples if you'd like to contribute something there.

**Comment by benanne** [comment]
Totally agreed that we went a bit overboard with the current examples, they should be simplified. As @ebenolson said we're already discussing this but I don't think anyone is planning to work on it right now. So your help would be very welcome :)

**Comment by f0k** [comment]
Well, I've been planning to simplify the example(s) last week, but had to postpone it. I'm again planning to do it this week, and fully agree with your second-to-last code block! The MNIST example will probably be converted to a flat file with as few indirections as possible anyway, focusing on an explanation of the library rather than giving a way to engineer your code for maximum reuse.

**Comment by buriy** [comment]
f0k, Also I haven't found non-trivial for newscomers to understand how to reuse the pipeline for macro batches.
Recompiling the pipeline for each macrobatch isn't a good solution.
Would be cool if you showed the data reloading in the example.
I mean, input_layer.set_value(...) instead of rebuild_network for consequent batches.

**Comment by f0k** [comment]
> Would be cool if you showed the data reloading in the example.

Have a look at #328, data transfer is a lot easier there. I don't use macro-batches as in my experience, performance benefits over just separately copying each batch to GPU were neglectible -- was it different for your case? It would be different if Theano allowed moving computation to a non-null stream and then using another non-null stream to update a shared variable (so data transfer time could be hidden completely), but for now this isn't possible.

**Comment by buriy** [comment]
f0k, I'm not an expert here -- I'm just evaluating the library and trying to get the maximum performance.
I'll now try changing macro-batch size and will look on the performance changes.
Unfortunately, documentation says nothing about performance.
I believe speed is a very important aspect for such libs -- otherwise  GPU computation wouldn't be needed at all!

**Comment by f0k** [comment]
Just a quick test on how the batch size influences the transfer time:
```python
In [1]: import numpy as np, theano
Using gpu device 0: GeForce GT 640

In [2]: X = np.random.randn(100000, 500).astype(np.float32)

In [3]: def loop(X, batchsize):
    batch = theano.shared(np.empty((batchsize,) + X.shape[1:], dtype=theano.config.floatX))
    for idx in range(len(X) // batchsize):
        batch.set_value(X[batchsize*idx:(batchsize+1)*idx], True)
   ...:         

In [4]: %timeit loop(X, 100)
1 loops, best of 3: 259 ms per loop

In [5]: %timeit loop(X, 1000)
100 loops, best of 3: 11.8 ms per loop

In [6]: %timeit loop(X, 10000)
100 loops, best of 3: 2.87 ms per loop

In [7]: %timeit loop(X, 100000)
10 loops, best of 3: 21.3 ms per loop
```

So yes, in relative terms, it is a lot faster to transfer 10,000 datapoints at once than to transfer them in mini-batches of 100 datapoints. But in absolute terms, you only save a fraction of a second per epoch. Feel free to investigate this further for more realistic data sizes (500 features is not a lot, and 100,000 datapoints is not a lot either), I'd be interested in more results. But for now I'm not convinced that macro-batches are a worthwhile route -- it makes the code more complicated and there's less GPU memory left for your model.

**Comment by buriy** [comment]
I have 100k samples in a macro-batch (~1GB). Internally I process them in mini-batches of 500 items. CPU to GPU data transfer takes about the same time as macro-batch calculation for me -- about 0.5 to 1 second. I'll switch to your proposed scheme later this week, and we'll see the impact on my use case.

**Comment by f0k** [comment]
Closed in #328. Github didn't understand my commit message.
